New syntax and extra

icecream17 opened the pull request:
This pr uses new javascript syntax

Also by using the new javascript syntax it could change functionality a bit

Also Deepcode noticed some things: https://www.deepcode.ai/app/gh/atom/atom/2beb93b0d7a0447150f52a9326d2c925e64254d2/_/dashboard and I've fixed some problems.

### Quantitative Performance Benefits

Well, the build doesn't even succeed yet (#22099). But I think updating will be better for performance

### Possible Drawbacks

Updating

### Verification Process

The build command doesn't work, but `atom-beta --test` does. And no errors, so that's great.

<details>
<summary>Note on commit 1</summary>
In the quicksort functions, or some other sort function, I think I changed "shift" to "pop". It's the same thing but faster.<br><br>
I changed the (Number) thing in some file to warn that binary, octal, and hex literals aren't supported now. (Because the old code doesn't check for this - it uses parseFloat.) Number is generally better than parseFloat but not that much. This doesn't have any affect on the code though.<br><br>
I randomly've gone through a whole bunch of files so this commit is messy<br><br>
Sometimes I wasn't sure if `??` could be used, and on those times, a few times I put it there anyways because any code that depends on the obscure exceptions should be fixed. But mostly I didn't change when I was unsure due to laziness<br><br>
CONTRIBUTING.MD changed, but only for consistency. This was caught by some extension in vscode.
</details>

<details>
<summary>Note on recent commits</summary>
Most of them just reference that Uzitech originally suggested these changes<br><br>
Some of them are lint fixes<br><br>
And then there's one commit that fixed a syntax error
</details>

TODO: 
- [ ] Revert quicksort function because the tests depend on them
- [ ] Linter fixes
- [ ] Split into multiple PRs
    - (I'll probably just close this one and make new ones from scratch)
    - [x] no-var (superseded by aminya's pr)
    - [ ] ??
    - [ ] ??=
    - [ ] other (maybe also function methods)

Edit: Depends on #21623 

-----
[View rendered CONTRIBUTING.md](https://github.com/icecream17/atom/blob/new-syntax/CONTRIBUTING.md)

icecream17 committed:
commit 1

icecream17 committed:
oops

UziTech commented on packages/deprecation-cop/lib/deprecation-cop-view.js:
```diff
@@ -137,7 +137,7 @@ export default class DeprecationCopView {
                   <div
                     className="list-item deprecation-message"
-                    innerHTML={marked(deprecation.getMessage())}
+                    dangerouslySetInnerHTML={marked(deprecation.getMessage())}
```
This is JSX for [etch](https://github.com/atom/etch) not react so I don't think `dangerouslySetInnerHTML` will work.

UziTech commented on packages/welcome/lib/welcome-view.js:
```diff
@@ -16,6 +16,9 @@ export default class WelcomeView {
     });
+
+    // See https://jsfiddle.net/reactjs/69z2wepo/
+    this.didChangeShowOnStartup = this.didChangeShowOnStartup.bind(this)
```
The `this` in that function refers to the element so this will break it.

UziTech commented on script/lib/verify-machine-requirements.js:
```diff
@@ -154,9 +154,6 @@ function verifyPython() {
     verifyBinary(
       path.join(process.env.SystemDrive || 'C:', 'Python27', 'python.exe')
     );
-    verifyBinary(
-      path.join(process.env.SystemDrive || 'C:', 'Python37', 'python.exe')
-    );
```
Why remove this?

icecream17 replied:
it's duplicate code, see lines before that

UziTech replied:
Nope `'Python27'` !== `'Python37'`

icecream17 replied:
Oh... I didn't see that

UziTech commented on src/cursor.js:
```diff
@@ -778,10 +780,7 @@ module.exports = class Cursor extends Model {
-      options && options.autoscroll != null
-        ? options.autoscroll
-        : this.isLastCursor();
+    const autoscroll = options.autoscroll ?? this.isLastCursor();
```
```suggestion
    const autoscroll = options?.autoscroll ?? this.isLastCursor();
```

UziTech commented on src/git-repository.js:
```diff
@@ -560,7 +563,6 @@ module.exports = class GitRepository {
     const repo = this.getRepo();
 
     const relativeProjectPaths =
-      this.project &&
       this.project
```
why remove this check? should the bottom line be `this.project?.`?

UziTech commented on src/main-process/file-recovery-service.js:
```diff
@@ -162,7 +162,7 @@ class RecoveryFile {
 async function tryStatFile(path) {
   return new Promise((resolve, reject) =>
-    fs.stat(path, (error, result) => resolve(error == null && result))
+    fs.stat(path, (error, result) => resolve(error ?? result))
```
These two lines are not equivalent.

`error == null && result` = `error == null ? result : false`
`error ?? result` = `error == null ? result : error`

in the second one `error` is a truthy value.

UziTech commented on src/core-uri-handlers.js:
```diff
@@ -3,7 +3,7 @@ const fs = require('fs-plus');
 // to a zero-based line or column number for the Atom API.
 function getLineColNumber(numStr) {
-  const num = parseInt(numStr || 0, 10);
+  const num = parseInt(numStr ?? 0, 10);
```
Not equivalent if `numStr` is `false`?

UziTech commented on src/text-editor.js:
```diff
@@ -1906,8 +1880,7 @@ module.exports = class TextEditor {
         while (
-          selection.start.row ===
-          (selections[0] != null ? selections[0].end.row : undefined)
+          selection.start.row === selections?.[0].end.row
```
```suggestion
          selection.start.row === selections[0]?.end.row
```

UziTech commented on src/text-editor.js:
```diff
@@ -4500,6 +4475,8 @@ module.exports = class TextEditor {
   //
   // * `grammar` {Grammar}
   setGrammar(grammar) {
+    console.warn('setGrammar is deprecated.')
```
```suggestion
    Grim.deprecate('setGrammar is deprecated.')
```

icecream17 committed:
thanks uzitech && use ??=

icecream17 committed:
Add back Python37

UziTech commented:
Nice work :100: It is great to see updated code.

FYI the tests might continue to fail until #22050 is merged. So don't bang your head if they are failing and you can't figure out why. lol

aminya commented on vendor/jasmine.js:
```diff
@@ -3,6 +3,7 @@
 // + var isCommonJS = typeof exports == "object";
 //
 // Modified method jasmine.WaitsForBlock.prototype.execute
+// Note: This file is also very outdated
```
You should not edit this file. It is under vendor

aminya commented on vendor/jasmine-jquery.js:
```diff
@@ -1,4 +1,6 @@
 'use strict'
+console.warn('\njasmine-jquery was last updated in May 28, 2017' +
```
You should not edit this file. It is under vendor

aminya reviewed (changes requested):
There are some issues with let/const usage. You have used `let` in many places where the variable is used only once. 

For these kinds of changes, you should use automatic tools that verify the changes using JavaScript AST which is more accurate than human eyes.

The whole `var` conversion can be done using jscodeshift. 

```
npm install -g jscodeshift
git clone https://github.com/cpojer/js-codemod.git
jscodeshift -t js-codemod/transforms/no-vars.js ./src
```

aminya commented on src/cursor.js:
```diff
@@ -611,11 +611,11 @@ module.exports = class Cursor extends Model {
 
     const ranges = this.editor.buffer.findAllInRangeSync(
-      options.wordRegex || this.wordRegExp(options),
+      options.wordRegex ?? this.wordRegExp(options),
```
Neither Node 12 nor Babel 5 support optional chaining or nullish coalescing

These changes require merging #21623

aminya commented on src/deserializer-manager.js:
```diff
@@ -63,7 +63,7 @@ module.exports = class DeserializerManager {
     if (deserializer) {
       let stateVersion =
-        (typeof state.get === 'function' && state.get('version')) ||
+        state.get?.('version') ??
```
Neither Node 12 nor Babel 5 support optional chaining or nullish coalescing

These changes require merging #21623

icecream17 committed:
fix obscure syntax error

I'm surprised this doesn't work

icecream17 committed:
I didn't delete it?

icecream17 committed:
linter errors 1-3 [skip ci]

icecream17 committed:
revert [skip ci]

icecream17 committed:
linter 5 [skip ci]

icecream17 committed:
lint 6, 7 [skip ci]

icecream17 committed:
Update jasmine-jquery.js

Update. Pun

icecream17 committed:
Undo vender changes

icecream17 committed:
undo vendor changes

icecream17 committed:
revert

icecream17 committed:
I thought I did these

Co-authored-by: Tony Brix <tony@brix.ninja>

icecream17 committed:
do all of suggestions again

Co-authored-by: Tony Brix <tony@brix.ninja>

icecream17 committed:
finish uzitech suggestions

I don't have access to a terminal right now so aminya's suggestions will happen later

aminya reviewed (changes requested):
@icecream17 This PR is already very hard to verify. Could you make separate PRs for editing individual changes (e.g. one PR for no-var conversion). Also, the commit messages are written very generically. Please be specific/verbose to help the reviewers.

aminya reviewed (changes requested):
I made a PR that supersedes the `var` conversion

https://github.com/atom/atom/pull/22103

icecream17 commented:
Closing (will make seperate prs in the future)
